### PR TITLE
No framework dependency injection

### PR DESCRIPTION
**Pull request by ch8n**
Hi @jmfayard 
I have added a small sample on how to do DI with Kotlin using no framework. please do the code review. 
Looking forward to your feedback.

Thanks!

Regards,
Chetan

### CODE REVIEW

**Comment by jmfayard** [comment]
@ch8n 

Please put your code in a package
```
@file:Suppress("PackageDirectoryMismatch")

package playground.di.native
```

And call the main fonction from [_main.kt](https://github.com/LouisCAD/kotlin-libraries-playground/blob/main/kotlin-jvm/src/main/kotlin/playground/_main.kt)

File: kotlin-jvm/src/main/kotlin/playground/NativeDI.kt [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+ * Native Dependency Injection: Pure Kotlin, no framework required Dependency Injection
+ * ideal for smaller-mid level projects which can be migrated easily to Dagger|Hilt like
+ * framewokr
+ *
+ * - [Blog](https://proandroiddev.com/hold-on-before-you-dagger-or-hilt-try-this-simple-di-f674c83ebeec)
+ * - [Github-Sample](https://github.com/ch8n/android-no-3rd-party-mvvm-example)
+ * - [Author](https://chetangupta.net)
+ */
+fun main() {
+
```
Comment (jmfayard) [inline kotlin-jvm/src/main/kotlin/playground/NativeDI.kt:16]: add

```
    println()
    println("# Native dependency injection in pure Kotlin")
```

File: kotlin-jvm/src/main/kotlin/playground/NativeDI.kt [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+                }
+            }
+        }
+        userCallback.invoke(user)
+    }
+}
+
+class SampleRepositoryImpl(
+        private val dataSourceLocal: DataSourceLocal,
+        private val dataSourceRemote: DataSourceRemote
```
Comment (LouisCAD) [inline kotlin-jvm/src/main/kotlin/playground/NativeDI.kt:117]: Please, fix indent to match Kotlin official style guide.

File: kotlin-jvm/src/main/kotlin/playground/NativeDI.kt [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+ */
+object Resolver {
+
+    fun provideViewModel(sampleRepository: SampleRepository): ViewModel {
+        return ViewModel(sampleRepository)
+    }
+
+    fun provideSampleRepository(localDataSource: DataSourceLocal, remoteDataSource: DataSourceRemote): SampleRepository {
+        return SampleRepositoryImpl(localDataSource, remoteDataSource)
+    }
```
Comment (LouisCAD) [inline kotlin-jvm/src/main/kotlin/playground/NativeDI.kt:43]: It'd be best to put these parameters on separate lines for readability (there's an IDE intention for that if you put the caret at the right place).

**Review by LouisCAD — CHANGES_REQUESTED** [review]

**Comment by ch8n** [comment]
> Please put your code in a package

should I make that package inside of playground or outside of it?

File: kotlin-jvm/src/main/kotlin/playground/NativeDI.kt (hunk 2 above)
Comment (ch8n) [reply kotlin-jvm/src/main/kotlin/playground/NativeDI.kt:117]: Hi @LouisCAD  I have set my studio with kotlin official style guide, it's showing the same result in auto-indentation and also checking on google show the same, can you tell me how I need to be formatted? something like
```
private val dataSourceLocal     : DataSourceLocal,
private val dataSourceRemote : DataSourceRemote
```

Comment (LouisCAD) [reply kotlin-jvm/src/main/kotlin/playground/NativeDI.kt:117]: @ch8n It should have only 4 spaces indent. Are you sure you saved the settings? Updating the code style and reformatting code with the IDE should do.

**Comment by jmfayard** [comment]
For the package, look how it's done for the other samples :)

File: kotlin-jvm/src/main/kotlin/playground/NativeDI.kt (hunk 3 above)
Comment (ch8n) [reply kotlin-jvm/src/main/kotlin/playground/NativeDI.kt:43]: > For the package, look how it's done for the other samples :)



> For the package, look how it's done for the other samples :)

Thanks, should I do `package playground.native.di` or `package playground.nativedi`

**Comment by ch8n** [comment]
> For the package, look how it's done for the other samples :)

Thanks , can you tell me which one to do ? `package playground.native.di` or `package playground.nativedi`

**Comment by LouisCAD** [comment]
Let's not have `native` in the name because it can be confused with Kotlin/Native and other uses of that word.

What about `playground.di.manual`?

File: kotlin-jvm/src/main/kotlin/playground/_main.kt [hunk 4]
```diff
@@ -22,6 +22,8 @@ fun main() {
     println()
     playground.exposed.main()
     println()
+    playground.native.di.main()
```
Comment (jmfayard) [inline kotlin-jvm/src/main/kotlin/playground/_main.kt:25]: Still wrong!

And please merge the `origin/main` branch

**Comment by jmfayard** [comment]
All good and merged, thanks Chetan! @ch8n

**Comment by ch8n** [comment]
Thanks guys! @jmfayard  @LouisCAD

**Comment by LouisCAD** [comment]
Thank **you**!

**Pull request merged**